### PR TITLE
RDRP-1004 imputation breakdowns

### PR DESCRIPTION
**Pull request by AnneONS**
## Pull Request submission

Created 3 types of breakdown checks: one for staging, which can optionally write a full data frame (while we debug), and a similar second one for imputation.

The third one is the existing construction one.

#### Closes or fixes

* Detail the ticket(s) you are closing with this PR
Closes #1004

#### Code

- [ ] **Code runs** The code runs on my machine and/or CDSW
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ ] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ ] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `requirements.txt`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

File: src/utils/breakdown_validation.py [hunk 1]
```diff
@@ -115,54 +148,224 @@ def greater_than_validation(
...
+        tuple(pd.DataFrame, pd.DataFrame): The filtered dataframe and the dataframe that
+          was removed.
+    """
+    validation_df = df[condition].copy()
+    removed_df = df[~condition].copy()
+    return validation_df, removed_df
+
+
+def run_breakdown_validation(
+    df: pd.DataFrame, config: dict, check: str
```
Comment (JenCheshire) [inline src/utils/breakdown_validation.py:340]: There's an additional arg now, that needs to be added when called in construction

Comment (AnneONS) [reply src/utils/breakdown_validation.py:340]: yep! got it, thanks

File: src/utils/breakdown_validation.py [hunk 2]
```diff
@@ -115,54 +148,224 @@ def greater_than_validation(
...
+    Raises:
+        ValueError: If any of the breakdown values do not sum to the total.
+
+    Returns:
+        None
+    """
+    to_check_dict = get_equality_dicts(config, "imputation")
+    check_results_dict, qa_df = get_breakdown_errors(df, to_check_dict)
+    # temp output of qa_df for debugging
+    if not qa_df.empty:
```
Comment (JenCheshire) [inline src/utils/breakdown_validation.py]: Were you going to comment the print statements out for the release?

File: src/utils/breakdown_validation.py [hunk 3]
```diff
@@ -115,54 +148,224 @@ def greater_than_validation(
...
+    Returns:
+        None
+    """
+    to_check_dict = get_equality_dicts(config)
+    df = replace_nulls_with_zero(df, to_check_dict)
+
+    check_results_dict, qa_df = get_breakdown_errors(df, to_check_dict)
+
+    # Note: this is a temporary implementation to show the QA output while we debug
+    if not qa_df.empty:
```
Comment (JenCheshire) [inline src/utils/breakdown_validation.py]: As above - comment out for release?

File: src/utils/breakdown_validation.py [hunk 4]
```diff
@@ -115,54 +148,224 @@ def greater_than_validation(
...
+    Returns:
+        None
+    """
+    for key, value in check_results_dict.items():
+        if not value.empty:
+            BreakdownValidationLogger.error(
+                f"Breakdown validation failed for {key} columns"
+            )
+            BreakdownValidationLogger.error(value)
     else:
```
Comment (JenCheshire) [inline src/utils/breakdown_validation.py]: Indentation of the else - it's currently logging the errors and the 'all vals are valid' msg

File: src/utils/breakdown_validation.py [hunk 5]
```diff
@@ -115,54 +148,224 @@ def greater_than_validation(
...
+
+    BreakdownValidationLogger.info(
+        f"There are {count} errors with the breakdown values"
+    )
     if not msg:
         BreakdownValidationLogger.info("All breakdown values are valid.")
     else:
         raise ValueError(msg)
 
     return df
```
Comment (JenCheshire) [inline src/utils/breakdown_validation.py]: I wonder if we comment out the value error (and returning the updated df), and just log the error message to the screen at this point?  Given we still have questions would it be safer for the moment

**Comment by JenCheshire** [comment]
Hi Anne, couple of comments - I also readded config as an arg for running in construction following merging in develop

File: src/utils/breakdown_validation.py (hunk 3 above)
Comment (AnneONS) [reply src/utils/breakdown_validation.py]: yes, thanks

**Review by JenCheshire — APPROVED** [review]

**Pull request merged**